File: integration/typescript/src/settlement_execution.ts
import type {
    ExecutedMarketMakerSettlement,
    MarketMakerSettlementExecutionPort,
    PreparedMarketMakerSettlement,
} from "./model.ts";

export type SettlementExecutionStage =
    | "preparation"
    | "simulation"
    | "submission"
    | "confirmation"
    | "reconciliation";

export class SettlementExecutionError extends Error {
    readonly stage: SettlementExecutionStage;
    readonly signature: string | null;

    constructor(
        stage: SettlementExecutionStage,
        message: string,
        signature: string | null = null,
    ) {
        super(message);
        this.name = "SettlementExecutionError";
        this.stage = stage;
        this.signature = signature;
    }
}

function executionErrorMessage(error: unknown): string {
    return error instanceof Error ? error.message : String(error);
}

export async function executePreparedMarketMakerSettlement<PreparedTransaction>(
    settlement: PreparedMarketMakerSettlement,
    executionPort: MarketMakerSettlementExecutionPort<PreparedTransaction>,
): Promise<ExecutedMarketMakerSettlement> {
    let transaction: PreparedTransaction;
    try {
        transaction = await executionPort.prepareTransaction(settlement);
    } catch (error: unknown) {
        throw new SettlementExecutionError(
            "preparation",
            executionErrorMessage(error),
        );
    }

    let simulation;
    try {
        simulation = await executionPort.simulate(transaction, settlement);
    } catch (error: unknown) {
        throw new SettlementExecutionError(
            "simulation",
            executionErrorMessage(error),
        );
    }
    if (!simulation.accepted) {
        throw new SettlementExecutionError("simulation", simulation.reason);
    }

    let signature: string;
    try {
        signature = await executionPort.submit(transaction, settlement);
    } catch (error: unknown) {
        throw new SettlementExecutionError(
            "submission",
            executionErrorMessage(error),
        );
    }
    if (signature.trim().length === 0) {
        throw new SettlementExecutionError(
            "submission",
            "Settlement submission returned an empty transaction signature",
        );
    }

    let confirmation;
    try {
        confirmation = await executionPort.confirm(signature, settlement);
    } catch (error: unknown) {
        throw new SettlementExecutionError(
            "confirmation",
            executionErrorMessage(error),
            signature,
        );
    }
    if (!confirmation.confirmed) {
        throw new SettlementExecutionError(
            "confirmation",
            confirmation.reason,
            signature,
        );
    }

    try {
        await executionPort.reconcile(signature, settlement);
    } catch (error: unknown) {
        throw new SettlementExecutionError(
            "reconciliation",
            executionErrorMessage(error),
            signature,
        );
    }

    return { action: settlement.action, signature };
}

File: integration/rust/src/settlement_execution.rs
use std::{error::Error, fmt};

use crate::{
    ExecutedMarketMakerSettlement,
    PreparedMarketMakerSettlement,
    SettlementConfirmationResult,
    SettlementSimulationResult,
};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum SettlementExecutionStage {
    Preparation,
    Simulation,
    Submission,
    Confirmation,
    Reconciliation,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SettlementExecutionError {
    pub stage: SettlementExecutionStage,
    pub signature: Option<String>,
    message: String,
}

impl SettlementExecutionError {
    pub fn new(
        stage: SettlementExecutionStage,
        message: impl Into<String>,
        signature: Option<String>,
    ) -> Self {
        Self {
            stage,
            signature,
            message: message.into(),
        }
    }
}

impl fmt::Display for SettlementExecutionError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl Error for SettlementExecutionError {}

pub trait MarketMakerSettlementExecutionPort {
    type Error: Error;
    type PreparedTransaction;

    fn prepare_transaction(
        &mut self,
        settlement: &PreparedMarketMakerSettlement,
    ) -> Result<Self::PreparedTransaction, Self::Error>;

    fn simulate(
        &mut self,
        transaction: &Self::PreparedTransaction,
        settlement: &PreparedMarketMakerSettlement,
    ) -> Result<SettlementSimulationResult, Self::Error>;

    fn submit(
        &mut self,
        transaction: &Self::PreparedTransaction,
        settlement: &PreparedMarketMakerSettlement,
    ) -> Result<String, Self::Error>;

    fn confirm(
        &mut self,
        signature: &str,
        settlement: &PreparedMarketMakerSettlement,
    ) -> Result<SettlementConfirmationResult, Self::Error>;

    fn reconcile(
        &mut self,
        signature: &str,
        settlement: &PreparedMarketMakerSettlement,
    ) -> Result<(), Self::Error>;
}

pub fn execute_prepared_market_maker_settlement<ExecutionPort>(
    settlement: &PreparedMarketMakerSettlement,
    execution_port: &mut ExecutionPort,
) -> Result<ExecutedMarketMakerSettlement, SettlementExecutionError>
where
    ExecutionPort: MarketMakerSettlementExecutionPort,
{
    let transaction = execution_port
        .prepare_transaction(settlement)
        .map_err(|error| {
            SettlementExecutionError::new(
                SettlementExecutionStage::Preparation,
                error.to_string(),
                None,
            )
        })?;

    let simulation = execution_port
        .simulate(&transaction, settlement)
        .map_err(|error| {
            SettlementExecutionError::new(
                SettlementExecutionStage::Simulation,
                error.to_string(),
                None,
            )
        })?;
    if let SettlementSimulationResult::Rejected { reason } = simulation {
        return Err(SettlementExecutionError::new(
            SettlementExecutionStage::Simulation,
            reason,
            None,
        ));
    }

    let signature = execution_port
        .submit(&transaction, settlement)
        .map_err(|error| {
            SettlementExecutionError::new(
                SettlementExecutionStage::Submission,
                error.to_string(),
                None,
            )
        })?;
    if signature.trim().is_empty() {
        return Err(SettlementExecutionError::new(
            SettlementExecutionStage::Submission,
            "Settlement submission returned an empty transaction signature",
            None,
        ));
    }

    let confirmation = execution_port
        .confirm(&signature, settlement)
        .map_err(|error| {
            SettlementExecutionError::new(
                SettlementExecutionStage::Confirmation,
                error.to_string(),
                Some(signature.clone()),
            )
        })?;
    if let SettlementConfirmationResult::Rejected { reason } = confirmation {
        return Err(SettlementExecutionError::new(
            SettlementExecutionStage::Confirmation,
            reason,
            Some(signature),
        ));
    }

    execution_port
        .reconcile(&signature, settlement)
        .map_err(|error| {
            SettlementExecutionError::new(
                SettlementExecutionStage::Reconciliation,
                error.to_string(),
                Some(signature.clone()),
            )
        })?;

    Ok(ExecutedMarketMakerSettlement {
        action: settlement.action,
        signature,
    })
}
